src/search/search-index/add.js

```diff
@@ -180,3 +180,32 @@ async function performIndexing(indexDoc) {
         console.error(err)
     }
 }
+
+async function indexingMetaConcurrent(type, pageId, timestamp) {
+    const reverseIndexDoc = await singleLookup(pageId)
+    if (type === 'visit') {
+        reverseIndexDoc.visits.add('visit/' + timestamp)
+    } else {
+        reverseIndexDoc.bookmarks.add('bookmark/' + timestamp)
+    }
+    await index.put(pageId, reverseIndexDoc)
+}
+
+/**
+ * Adds a new meta index (bookmark or visit) entry.
+ * NOTE: Assumes the existence of indexed `pageId`.
+ *
+ * @param {'bookmark'|'visit'} type Type of meta event.
+ * @param {string} timestamp Timestamp of meta event.
+ * @param {string} pageId ID of page doc to associate with.
+ */
+export const addMetaConcurrent = (type, timestamp, pageId) =>
+    new Promise((resolve, reject) =>
+        indexQueue.push(() => {
+            indexingMetaConcurrent(type, pageId, timestamp)
```

Comment:
As both the `indexingMetaConcurrent` (again, this needs renaming; maybe `addMetaToReversePage`?) and `index.put` calls in here are async, there will be cases where the Promise on L203 gets resolved before `indexingMetaConcurrent` is finished. To fix this, you could just do them sequentially, maybe just adding another `.then` like:

```js
indexingMetaConcurrent(...)
  .then(() => index.put(...))
  .then(resolve)
  .catch(reject)
```